app/controllers/application_controller.rb

```diff
@@ -5,6 +5,152 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
   end
 
-end
\ No newline at end of file
+  get '/' do
+
+    erb :index
+  end
+
+  get '/signup' do
+    if logged_in?
+      redirect "/tweets"
+    end
+    erb :'/users/create_user'
+  end
+
+  post '/signup' do
+    if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect "/signup"
+    else
+
+      @user = User.create(params)
+      session[:user_id] = @user.id
+      redirect "/tweets"
+    end
+
+    erb :'/users/create_user'
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect "/tweets"
+    end
+    erb :'/users/login'
+  end
+
+  post '/login' do
+    @user = User.find_by_username(params[:username])
+
+    if User.all.include?(@user) && @user.authenticate(params[:password])
+      session[:user_id] = @user.id
+
+      redirect "/tweets"
+    end
+
+    erb :'/users/login'
+  end
+
+  get '/logout' do
+    if logged_in?
+      session.clear
+      redirect "/login"
+    end
+    redirect "/tweets"
+  end
+
+  get '/tweets' do
+    @tweets = Tweet.all
+
+    if !logged_in?
+      redirect "/login"
+    end
+    erb :'/tweets/tweets'
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+
+    erb :'/users/show'
+  end
+
+  get '/tweets/new' do
+    if !logged_in?
+      redirect "/login"
+    end
+
+    erb :'/tweets/create_tweet'
+  end
+
+  post '/tweets' do
+    if params[:content] == ""
+      redirect "/tweets/new"
+    elsif logged_in?
+      @tweet = Tweet.create(content: params[:content], user_id: session[:user_id])
+
+      redirect "/tweets"
+    end
+  end
+
+  get '/tweets/:id' do
+    @tweet = Tweet.find(params[:id])
+
+    if logged_in?
+      erb :'/tweets/show_tweet'
+    elsif !logged_in?
+      redirect "/login"
+    end
+  end
+
+  get '/tweets/:id/edit' do
+
+    @tweet = Tweet.find_by_id(params[:id])
+
+    if logged_in? && @tweet.user_id ==  current_user.id
+  erb :'/tweets/edit_tweet'
+    elsif !logged_in?
+      redirect "/login"
+    end
+  end
+
+  patch '/tweets/:id' do
+
+
+    @tweet = Tweet.find(params[:id])
+    if logged_in? && @tweet.user_id == current_user.id
+      if params[:content] != "" && params[:content] != nil
+        @tweet.content = params[:content]
+        @tweet.save
+      elsif params[:content] == ""
+        redirect "/tweets/#{@tweet.id}/edit"
+      end
+    elsif !logged_in?
+      redirect "/login"
+    end
+  end
+
+  delete '/tweets/:id/delete' do
+    @tweet = Tweet.find(params[:id])
+    if !logged_in?
+      redirect "/login"
+    elsif logged_in? && @tweet.user_id == current_user.id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.